### ISSUE TITLE
Issue importing attributes of type akeneo_reference_entity

### ISSUE THREAD

**Issue by sarramegnag** [synolia/SyliusAkeneoPlugin#71]
Hi,

We're having problems importing products which references to an `akeneo_reference_entity`. In order to do this, firstly we need to import attributes in `sylius_product_option` table and their values in `sylius_product_option_value` with the `akeneo:import:attributes` command.

The previous command retrieves attributes and when it does import them as options, this line unfortunatly ignores everything which is not `pim_catalog_simpleselect` / `select` (`SelectAttributeTypeMatcher`) or `pim_catalog_multiselect` / `multiselect` (`MultiSelectAttributeTypeMatcher`) :
https://github.com/synolia/SyliusAkeneoPlugin/blob/v2.1.0/src/Task/AttributeOption/RetrieveOptionsTask.php#L55
Therefore my resource with type `akeneo_reference_entity` (`ReferenceEntityAttributeTypeMatcher`) is skipped and values in sylius_product_option_value are not created.

Do you have any information why this is not imported ?

Best regards.

This is part of my https://*****.cloud.akeneo.com/api/rest/v1/attributes call :
```
// ...
      {
        "_links": {
          "self": {
            "href": "https:\/\/*****.cloud.akeneo.com\/api\/rest\/v1\/attributes\/gce_color_variant"
          }
        },
        "code": "gce_color_variant",
        "type": "akeneo_reference_entity",
        "group": "axe_declinaison",
        "unique": false,
        "useable_as_grid_filter": false,
        "allowed_extensions": [],
        "metric_family": null,
        "default_metric_unit": null,
        "reference_data_name": "coloris_gce",
        "available_locales": [],
        "max_characters": null,
        "validation_rule": null,
        "validation_regexp": null,
        "wysiwyg_enabled": null,
        "number_min": null,
        "number_max": null,
        "decimals_allowed": null,
        "negative_allowed": null,
        "date_min": null,
        "date_max": null,
        "max_file_size": null,
        "minimum_input_length": null,
        "sort_order": 0,
        "localizable": false,
        "scopable": false,
        "labels": {
          "fr_FR": "Coloris"
        },
        "guidelines": {},
        "auto_option_sorting": null,
        "is_read_only": false,
        "default_value": null,
        "group_labels": {
          "fr_FR": "Axe déclinaison"
        }
      },
// ...
```

And call to https://*****.cloud.akeneo.com/api/rest/v1/reference-entities/coloris_gce/records :
```
// ...
      {
        "_links": {
          "self": {
            "href": "https:\/\/*****.cloud.akeneo.com\/api\/rest\/v1\/reference-entities\/coloris_gce\/records\/t01_1002_bordeau"
          }
        },
        "code": "t01_1002_bordeau",
        "values": {
          "label": [
            {
              "locale": "fr_FR",
              "channel": null,
              "data": "Bordeaux"
            }
          ],
          "image": [
            {
              // ...
            }
          ],
          "gce_supplier_name": [
            {
              "locale": null,
              "channel": null,
              "data": "John Doe"
            }
          ],
          "gce_supplier_internal_fabric_reference": [
            {
              "locale": null,
              "channel": null,
              "data": "Tissu bonnie"
            }
          ],
          "pim_main_color": [
            {
              "locale": null,
              "channel": null,
              "data": [
                "rouge"
              ]
            }
          ]
        }
      },
// ...
```

**Comment by TheGrimmChester** [comment]
Hello @sarramegnag,

I think you have to register the reference entity json processor in your services and enable the enterprise feature from the akeneo api configuration section in the Sylius BO.

```yaml
Synolia\SyliusAkeneoPlugin\Processor\ProductAttributeValue\ReferenceEntity\JsonReferenceEntityAttributeValueProcessor:
    tags:
        - { name: !php/const Synolia\SyliusAkeneoPlugin\Processor\ProductAttributeValue\ReferenceEntity\ReferenceEntityAttributeValueProcessorInterface::TAG_ID }
```

**Comment by sarramegnag** [comment]
Hi @TheGrimmChester,

First, thanks for your message !

The configuration already uses the enterprise feature from Akeneo API.

I have dumped in both files and what I've got is:

CreateUpdateEntityTask.php on line 128
```php
dump($resource['type']); // "akeneo_reference_entity"
dump($attributeType); // Synolia\SyliusAkeneoPlugin\TypeMatcher\Attribute\ReferenceEntityAttributeTypeMatcher {#5332}
dump($attributeType instanceof ReferenceEntityAttributeTypeMatcher); // true
dump($apiConfiguration->isEnterprise()); // true
```

and I dumped in `Synolia\SyliusAkeneoPlugin\Processor\ProductAttributeValue\ReferenceEntity\JsonReferenceEntityAttributeValueProcessor::getValue()` but nothing showed up.

Line has been created in `sylius_product_attribute`, but I have `a:0:{}` in configuration column (which I'm not sure is normal):

| id | code | type | storage\_type | configuration | created\_at | updated\_at | position | custom\_imported\_from\_pim | custom\_show\_on\_product\_card | custom\_show\_on\_comparator | custom\_use\_as\_facet | custom\_export\_to\_fred\_hopper |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 5 | gce\_color\_variant | reference\_entity | json | a:0:{} | 2021-07-15 16:10:28 | 2021-07-15 16:10:28 | 2 | 0 | 1 | 0 | 0 | 0 |

Line has been created in `sylius_product_option`:

| id | code | created\_at | updated\_at | position |
| :--- | :--- | :--- | :--- | :--- |
| 1 | gce\_color\_variant | 2021-07-15 16:11:20 | 2021-07-15 16:11:20 | 0 |
| 2 | gce\_dimension\_width\_length\_variant | 2021-07-15 16:11:20 | 2021-07-15 16:11:20 | 1 |
| 3 | gce\_sofa\_design\_variant | 2021-07-15 16:11:20 | 2021-07-15 16:11:20 | 2 |
| 4 | test\_couleur | 2021-07-15 16:11:20 | 2021-07-15 16:11:20 | 3 |
| 5 | test\_finition\_rideaux | 2021-07-15 16:11:20 | 2021-07-15 16:11:20 | 4 |


I have got nothing in `sylius_product_option_value`, where I should have my different colors from call to `https://*****.cloud.akeneo.com/api/rest/v1/reference-entities/coloris_gce/records` (response in first post).

Any idea ?

Regards.

**Comment by TheGrimmChester** [comment]
@sarramegnag can you debug what is passed inside the method src/Manager/ProductOptionManager.php:156 it's where all option values are handled.

**Comment by sarramegnag** [comment]
On dumping at the very beginning of method updateProductOptionValues(), here what I have respectively for ProductOptionInterface $productOption and AttributeInterface $attribute:

```
App\Entity\Sylius\Product\ProductOption {#17481 ▼
  #id: null
  #code: "gce_color_variant"
  #position: null
  #values: Doctrine\Common\CollectionsArrayCollection {#17489 ▶}
  #createdAt: DateTime @1626707230 {#17493 ▶}
  #updatedAt: DateTime @1626707230 {#16960 ▶}
  #translations: Doctrine\Common\CollectionsArrayCollection {#17487 ▶}
  #translationsCache: array:1 [▶]
  #currentLocale: "fr_FR"
  #currentTranslation: null
  #fallbackLocale: "fr_FR"
}

App\Entity\Sylius\Product\ProductAttribute {#6835 ▼
  #importedFromPim: false
  #showOnProductCard: false
  #showOnComparator: false
  #useAsFacet: false
  #exportToFredHopper: false
  #id: 5
  #code: "gce_color_variant"
  #type: "reference_entity"
  #configuration: []
  #storageType: "json"
  #position: 2
  #createdAt: DateTime @1626707147 {#6839 ▶}
  #updatedAt: DateTime @1626707147 {#6849 ▶}
  #translations: Doctrine\ORMPersistentCollection {#6853 ▶}
  #translationsCache: array:1 [▶]
  #currentLocale: "fr_FR"
  #currentTranslation: null
  #fallbackLocale: "fr_FR"
}
```

It then gets ignored by next line:
```php
if ($attribute->getType() !== SelectAttributeType::TYPE) {
  return;
}
```

That's the point I don't understand.

**Comment by TheGrimmChester** [comment]
Hi @sarramegnag,

We only import reference entities as json for performance reason so it's fine to not pass inside this method. What is important to have (after all imports) is entries inside the table sylius_product_attribute_value for your reference entity attribute.

You should see something like:

```
# id, product_id, attribute_id, text_value, boolean_value, integer_value, float_value, datetime_value, date_value, json_value, locale_code
'364', '2', '88', NULL, NULL, NULL, NULL, NULL, NULL, 'YOUR_REFERNCE_ENTITY_JSON_VALUE', 'fr_FR'
```
It will be displayed on the product page (FO & BO) and you will probably need to customize the template to render the information.

**Comment by sarramegnag** [comment]
Hi @TheGrimmChester ,

Thanks again for your support. 

For now, the table `sylius_product_attribute_value` is empty. I'll try to investigate why.

Any idea ?

**Comment by sarramegnag** [comment]
I continued investigating @TheGrimmChester and I would like to clarify something.

In my case, the Akeneo attribute is a variation axe, thus I think it should be stored in `sylius_product_option` to handle variation of a product helped by `sylius_product_variant` table.

So, in my case, it should be a `sylius_product_option`, but since its type is reference_entity in `sylius_product_attribute` table, it's not loaded as an option because of the `if ($attribute->getType() !== SelectAttributeType::TYPE) {` line.

For example, I have another attribute called `gce_manufacturer` which is a reference_entity too, but not a variation axe, and its values are properly stored in `sylius_product_attribute_value` table.

sylius_product_attribute:
| id | code | type | storage\_type | configuration | created\_at | updated\_at | position | custom\_imported\_from\_pim | custom\_show\_on\_product\_card | custom\_show\_on\_comparator | custom\_use\_as\_facet | custom\_export\_to\_fred\_hopper |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 14 | gce\_manufacturer | reference\_entity | json | a:0:{} | 2021-07-20 08:21:46 | 2021-07-20 08:21:46 | 8 | 0 | 0 | 0 | 0 | 0 |

sylius_product_attribute_value:
| id | product\_id | attribute\_id | text\_value | boolean\_value | integer\_value | float\_value | datetime\_value | date\_value | json\_value | locale\_code |
| :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- | :--- |
| 1599 | 60 | 14 | NULL | NULL | NULL | NULL | NULL | NULL | {"code":"0696\_ENO","attributes":{"label":"ENO","image":"b\\/4\\/0\\/3\\/b403129b7bd3412a02299cbee343c952d1bc5b73\_Eno\_logo.jpg","gce\_brand\_number":"0696"}} | fr\_FR |

I also have an attribute called `gce_dimension_width_length_variant` of type `pim_catalog_simpleselect` which is a variation axe and it's properly created in `sylius_product_option` and `sylius_product_option_value` tables.

sylius_product_option:
| id | code | created\_at | updated\_at | position |
| :--- | :--- | :--- | :--- | :--- |
| 2 | gce\_dimension\_width\_length\_variant | 2021-07-20 08:22:43 | 2021-07-20 08:22:43 | 1 |

sylius_product_option_value:
| id | option\_id | code |
| :--- | :--- | :--- |
| 1 | 2 | gce\_dimension\_width\_length\_variant\_akeneo-test\_100\_x\_190\_cm |
| 2 | 2 | gce\_dimension\_width\_length\_variant\_akeneo-test\_100\_x\_200\_cm |

In conclusion, since my Akeneo attribute `gce_variant_color` is an akeneo_reference_entity AND it's a variation axe, it looks like it's not handled by the plugin.

Do you think I did something wrong and then it should be handled by the plugin ?

**Comment by TheGrimmChester** [comment]
@sarramegnag you did nothing wrong. I wasn't understanding that you used a reference entity as variation axe. 
This case is not handled by the plugin at this time because reference entities are complex to use. If you can achieve it, your PR will be welcome!

**Comment by sarramegnag** [comment]
Thank you @TheGrimmChester , now I know I have to do it myself :)

I will close the issue, but maybe I will be able to do a PR later. Who knows ?

**Comment by oallain** [comment]
Thanks for feedback @sarramegnag 👍🏼 
Your PR is welcome if you have the opportunity to do it 🚀

**Comment by TheGrimmChester** [comment]
@sarramegnag I started to develop this feature. This is still in WIP, but it's available here https://github.com/synolia/SyliusAkeneoPlugin/pull/73/files

**Comment by sarramegnag** [comment]
Well... I'm not sure I will be able to make a PR soon, so I can at least give you my little hack if you need inspiration :)
But I will gladly update the bundle and get rid of my hack if you manage to make it work !

```php
<?php

namespace App\Command\Import\EventSubscriber;

use Akeneo\Pim\ApiClient\Search\SearchBuilder;
use App\Command\Import\Pagination\ResourceCursor;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Synolia\SyliusAkeneoPlugin\Event\AfterTaskEvent;
use Synolia\SyliusAkeneoPlugin\Payload\Option\OptionsPayload;
use Synolia\SyliusAkeneoPlugin\Task\AttributeOption\RetrieveOptionsTask;

class RetrieveReferenceEntitiesRecordsTask implements EventSubscriberInterface
{
    public static function getSubscribedEvents(): array
    {
        return [
            AfterTaskEvent ::class => 'retrieveReferenceEntitiesRecords',
        ];
    }

    public function retrieveReferenceEntitiesRecords(AfterTaskEvent $event): void
    {
        if ($event->getTask() !== RetrieveOptionsTask::class) {
            return;
        }

        /** @var OptionsPayload $payload */
        $payload = $event->getPayload();
        $resources = $payload->getResources();

        $searchBuilder = (new SearchBuilder())
            ->addFilter('type', 'IN', ['akeneo_reference_entity', 'akeneo_reference_entity_collection'])
        ;

        $referenceEntities = $payload->getAkeneoPimClient()->getAttributeApi()->all(100, [
            'search' => $searchBuilder->getFilters(),
        ]);

        foreach ($referenceEntities as $referenceEntity) {
            $records = $payload->getAkeneoPimClient()->getReferenceEntityRecordApi()->all(
                $referenceEntity['reference_data_name']
            );

            $items = array_map(static function (array $record): array {
                return [
                    'code' => $record['code'],
                    'labels' => array_reduce(
                        $record['values']['label'],
                        static function (array $carry, array $label): array {
                            $carry[$label['locale']] = $label['data'];

                            return $carry;
                        },
                        []
                    )
                ];
            }, iterator_to_array($records));

            $resourceCursor = new ResourceCursor($items);

            $resources[$referenceEntity['code']] = [
                'isMultiple' => $referenceEntity['type'] === 'akeneo_reference_entity_collection',
                'resources' => $resourceCursor,
            ];
        }

        $payload->setResources($resources);
    }
}

```

**Comment by TheGrimmChester** [comment]
I managed to import reference entity record codes as option values ans now I need to bind the option to the variant and make the code cleaner
